src/lib.rs

```diff
@@ -854,6 +876,29 @@ impl<T, const N: usize> FromFn<T> for [T; N] {
     {
         core::array::from_fn(cb)
     }
+
+    // TODO(tarcieri): use `array::try_from_fn` when stable
+    fn try_from_fn<E, F>(mut cb: F) -> Result<Self, E>
+    where
+        F: FnMut(usize) -> Result<T, E>,
+    {
+        // SAFETY: an array of `MaybeUninit`s is always valid.
+        let mut array: [MaybeUninit<T>; N] = unsafe { MaybeUninit::uninit().assume_init() };
+
+        #[allow(clippy::needless_range_loop)]
+        for i in 0..N {
+            array[i].write(cb(i)?);
+        }
```

Comment:
This is still not quite correct. If `cb` panics in the middle, you will not call `Drop` on previously initialized items.

Reply:
I guess `core` uses a drop guard for constructing the array which keeps track of the number of initialized items and also impls `Drop`, so it can take care of dropping them if the partially constructed array gets dropped: https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/src/core/array/mod.rs.html#812-830

Reply:
Yes and we need something similar. Alternatively, we need to somehow limit `T`.

Reply:
I'll mimic the approach `core` uses

Reply:
Done in [2d8591f](https://github.com/RustCrypto/hybrid-array/pull/2/commits/2d8591f32b036a691d60bebdf9abb5661931f02d)